iree/hal/local/loaders/embedded_library_loader.c

```diff
@@ -0,0 +1,296 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "iree/hal/local/loaders/embedded_library_loader.h"
+
+#include "iree/base/target_platform.h"
+#include "iree/base/tracing.h"
+#include "iree/hal/local/elf/elf_module.h"
+#include "iree/hal/local/local_executable.h"
+
+//===----------------------------------------------------------------------===//
+// iree_hal_elf_executable_t
+//===----------------------------------------------------------------------===//
+
+typedef struct {
+  iree_hal_local_executable_t base;
+
+  // Loaded ELF module.
+  iree_elf_module_t module;
+
+  // Name used for the file field in tracy and debuggers.
+  iree_string_view_t identifier;
+
+  // Queried metadata from the library.
+  union {
+    const iree_hal_executable_library_header_t** header;
+    const iree_hal_executable_library_v0_t* v0;
+  } library;
+} iree_hal_elf_executable_t;
+
+extern const iree_hal_local_executable_vtable_t iree_hal_elf_executable_vtable;
+
+static iree_status_t iree_hal_elf_executable_query_library(
+    iree_hal_elf_executable_t* executable) {
+  // Get the exported symbol used to get the library metadata.
+  iree_hal_executable_library_query_fn_t query_fn = NULL;
+  IREE_RETURN_IF_ERROR(iree_elf_module_lookup_export(
+      &executable->module, IREE_HAL_EXECUTABLE_LIBRARY_EXPORT_NAME,
+      (void**)&query_fn));
+
+  // Query for a compatible version of the library.
+  executable->library.header =
+      (const iree_hal_executable_library_header_t**)iree_elf_call_p_i(
+          query_fn, IREE_HAL_EXECUTABLE_LIBRARY_LATEST_VERSION);
+  if (!executable->library.header) {
+    return iree_make_status(
+        IREE_STATUS_FAILED_PRECONDITION,
+        "executable does not support this version of the runtime (%d)",
+        IREE_HAL_EXECUTABLE_LIBRARY_LATEST_VERSION);
+  }
+  const iree_hal_executable_library_header_t* header =
+      *executable->library.header;
+
+  // Ensure that if the library is built for a particular sanitizer that we also
+  // were compiled with that sanitizer enabled.
+  switch (header->sanitizer) {
```

Comment:
It seems like ultimately, whether the executable was compiled for a sanitizer should be a multi-versioning flag that we select an executable based on (i.e. same as detecting a variant based on cpu features).

Reply:
Changed this to always fail if any sanitizers were compiled in - getting compilerrt stable across platforms is .... not worth it :P